File: app/src/main/java/com/roana/app/CorridorStateMachine.kt
package com.roana.app

import com.roana.app.CorridorPlanner.CorridorCommand
import com.roana.app.CorridorPlanner.CorridorDecision

class CorridorStateMachine(
    private val confirmationsRequired: Int = CorridorContract.DEFAULT_CONFIRMATIONS_REQUIRED,
) {
    private var currentCommand = CorridorCommand.STOP
    private var pendingCommand: CorridorCommand? = null
    private var pendingCount = 0

    init {
        require(confirmationsRequired > 0) { "confirmationsRequired must be positive" }
    }

    fun update(decision: CorridorDecision): CorridorState {
        val proposedCommand = if (decision.requiresEmergencyStop()) {
            CorridorCommand.STOP
        } else {
            decision.command
        }
        if (proposedCommand == currentCommand) {
            pendingCommand = null
            pendingCount = 0
            return state(decision, changed = false)
        }

        if (proposedCommand == CorridorCommand.STOP) {
            currentCommand = CorridorCommand.STOP
            pendingCommand = null
            pendingCount = 0
            return state(decision, changed = true)
        }

        if (pendingCommand == proposedCommand) {
            pendingCount += 1
        } else {
            pendingCommand = proposedCommand
            pendingCount = 1
        }

        if (pendingCount >= confirmationsRequired) {
            currentCommand = proposedCommand
            pendingCommand = null
            pendingCount = 0
            return state(decision, changed = true)
        }

        return state(decision, changed = false)
    }

    private fun state(decision: CorridorDecision, changed: Boolean): CorridorState =
        CorridorState(
            command = currentCommand,
            sourceDecision = decision,
            pendingCommand = pendingCommand,
            pendingCount = pendingCount,
            changed = changed,
        )

    private fun CorridorDecision.requiresEmergencyStop(): Boolean =
        CorridorContract.Reason.isEmergencyStop(reason)

    data class CorridorState(
        val command: CorridorCommand,
        val sourceDecision: CorridorDecision,
        val pendingCommand: CorridorCommand?,
        val pendingCount: Int,
        val changed: Boolean,
    )
}

File: ios/Roana/Roana/Corridor/CorridorStateMachine.swift
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2026 The Roana Authors.

import Foundation

final class CorridorStateMachine {
    private let confirmationsRequired: Int
    private var currentCommand = CorridorCommand.stop
    private var pendingCommand: CorridorCommand?
    private var pendingCount = 0

    init(confirmationsRequired: Int = 3) {
        precondition(confirmationsRequired > 0, "confirmationsRequired must be positive")
        self.confirmationsRequired = confirmationsRequired
    }

    func update(decision: CorridorDecision) -> CorridorState {
        let proposedCommand = decision.requiresEmergencyStop ? CorridorCommand.stop : decision.command
        if proposedCommand == currentCommand {
            pendingCommand = nil
            pendingCount = 0
            return state(decision: decision, changed: false)
        }

        if proposedCommand == .stop {
            currentCommand = .stop
            pendingCommand = nil
            pendingCount = 0
            return state(decision: decision, changed: true)
        }

        if pendingCommand == proposedCommand {
            pendingCount += 1
        } else {
            pendingCommand = proposedCommand
            pendingCount = 1
        }

        if pendingCount >= confirmationsRequired {
            currentCommand = proposedCommand
            pendingCommand = nil
            pendingCount = 0
            return state(decision: decision, changed: true)
        }

        return state(decision: decision, changed: false)
    }

    private func state(decision: CorridorDecision, changed: Bool) -> CorridorState {
        CorridorState(
            command: currentCommand,
            sourceDecision: decision,
            pendingCommand: pendingCommand,
            pendingCount: pendingCount,
            changed: changed,
        )
    }
}

struct CorridorState: Equatable {
    let command: CorridorCommand
    let sourceDecision: CorridorDecision
    let pendingCommand: CorridorCommand?
    let pendingCount: Int
    let changed: Bool
}

private extension CorridorDecision {
    var requiresEmergencyStop: Bool {
        reason == "frame_loss" || reason == "low_confidence"
    }
}
